crates/holochain/src/conductor/interface/websocket.rs

```diff
@@ -165,58 +203,111 @@ async fn recv_incoming_admin_msgs<A: InterfaceApi>(api: A, rx_from_iface: Websoc
     rx_from_cell: broadcast::Receiver<Signal>,
     tx_to_iface: WebsocketSender,
 ) {
     use futures::stream::StreamExt;
 
-    trace!("CONNECTION: {}", rx_from_iface.remote_addr());
+    trace!("CONNECTION: {}", rx_from_iface.peer_addr());
 
     let rx_from_cell = futures::stream::unfold(rx_from_cell, |mut rx_from_cell| async move {
-        if let Ok(item) = rx_from_cell.recv().await {
-            Some((item, rx_from_cell))
-        } else {
-            None
-        }
-    });
-
-    tokio::task::spawn(rx_from_cell.for_each_concurrent(4096, move |signal| {
-        let mut tx_to_iface = tx_to_iface.clone();
-        async move {
-            trace!(msg = "Sending signal!", ?signal);
-            if let Err(err) = async move {
-                let bytes = SerializedBytes::try_from(signal)?;
-                tx_to_iface.signal(bytes).await?;
-                InterfaceResult::Ok(())
-            }
-            .await
-            {
-                error!(?err, "error emitting signal");
+        loop {
+            match rx_from_cell.recv().await {
+                // We missed some signals, but the channel is still open
+                Err(tokio::sync::broadcast::error::RecvError::Lagged(_)) => {
+                    warn!("SignalChannelOverloaded");
+                    continue;
+                }
+                Ok(item) => return Some((item, rx_from_cell)),
+                _ => {
+                    warn!("SignalChannelClosed");
+                    return None;
+                }
             }
         }
-    }));
+    });
 
-    tokio::task::spawn(rx_from_iface.for_each_concurrent(4096, move |msg| {
-        let api = api.clone();
-        async move {
-            if let Err(err) = handle_incoming_message(msg, api).await {
-                error!(?err, "error handling websocket message");
+    task_list
+        .0
+        .push(tokio::task::spawn(rx_from_cell.for_each_concurrent(
+            CONCURRENCY_COUNT,
+            move |signal| {
+                let tx_to_iface = tx_to_iface.clone();
+                async move {
+                    trace!(msg = "Sending signal!", ?signal);
+                    if let Err(err) = async move {
+                        tx_to_iface.signal(signal).await?;
+                        InterfaceResult::Ok(())
+                    }
+                    .await
+                    {
+                        error!(?err, "error emitting signal");
+                    }
+                }
+            },
+        )));
+
+    let rx_from_iface =
+        futures::stream::unfold(rx_from_iface, move |mut rx_from_iface| async move {
+            match rx_from_iface.recv().await {
+                Ok(r) => Some((r, rx_from_iface)),
+                Err(err) => {
+                    info!(?err);
+                    None
+                }
             }
-        }
-    }));
+        });
+
+    task_list
+        .0
+        .push(tokio::task::spawn(rx_from_iface.for_each_concurrent(
```

Comment:
There are a few places where `for_each_concurrent` has been used which I'm not too concerned about, but here it is one of the places that we could potentially see backups when the concurrency limit is reached and messages are taking a bit of time to process. Could we track that behavior here like we do in a few other places with `for_each_concurrent`?

e.g. https://github.com/holochain/holochain/blob/7019aa187a0c0fd6dae44394b0ddad4a52e8062c/crates/holochain/src/conductor/conductor.rs#L3056

This isn't a new thing because this code has been moved and just looks new so I'd be fine with a TODO. But it's something I'm keen to track unless you think it's not so useful here! :)

Reply:
Yep, metrics could be useful here. Went with the todo for now, though : )